core/conn/unixodbc/odbc/odbcclient/unixcli/cli/ctosqlconv.cpp

```diff
@@ -1036,2500 +1166,3031 @@ unsigned long ODBC::ConvertCToSQL(SQLINTEGER	ODBCAppVersion,
-					break;
-				case SQLTYPECODE_LARGEINT:
-					tempVal64 = (__int64)dTmp;
-					DataPtr = &tempVal64;
-					DataLen = sizeof(tempVal64);
-					break;
-				default:
-					return IDS_07_006;
-				}
-			}
-			else
-			{
-				if( targetUnsigned && negative )
-					return IDS_22_003_02;	//negValue in unsigned column
-
-				if (targetScale)
-				{
-					for (i = 0,tempVal64 = 1; i < targetScale ;  i++)
-						tempVal64 *= 10;
-					tempVal64 = tempVal64 * integralPart;
-					decimalDigits = 0;
-					if (decimalPart > 0)
-						decimalDigits = getDigitCount(decimalPart);
-					scaleOffset = 0;
-					if (leadZeros < targetScale)
-					  scaleOffset = targetScale - decimalDigits - leadZeros;
-					if (scaleOffset < 0)
-					{
-//NUMERIC_VALUE_OUT_OF_RANGE_ERROR
-						return IDS_22_003;
-					}
-
-					for (i =0, tempScaleVal64 = decimalPart ; i < scaleOffset ; i++)
-						tempScaleVal64 *= 10;
-					tempVal64 += tempScaleVal64;
-				}
-				else
-				{
-//NUMERIC_DATA_TRUNCATED_ERROR
-					if (decimalPart != 0)
-						retCode = IDS_01_S07;
-					tempVal64 = integralPart;
-				}
-				if (negative)
-					tempVal64 = -tempVal64;
-
-				switch( SQLDataType )
-				{
-				case SQLTYPECODE_SMALLINT_UNSIGNED:
-					if (tempVal64 < 0)
-					       return IDS_22_003_02;
-					if ((USHORT)tempVal64 > USHRT_MAX)
-						return IDS_22_003;
-					usTmp = (USHORT)tempVal64;
-					if  (tempVal64 != usTmp)
-						retCode = IDS_01_S07;
-					DataPtr = &usTmp;
-					DataLen = sizeof(USHORT);
-					break;
-				case SQLTYPECODE_SMALLINT:
-					if (tempVal64 < SHRT_MIN || tempVal64 > SHRT_MAX)
-						return IDS_22_003;
-					sTmp = (SHORT)tempVal64;
-					if  (tempVal64 != sTmp)
-						retCode = IDS_01_S07;
-					DataPtr = &sTmp;
-					DataLen = sizeof(sTmp);
-					break;
-				case SQLTYPECODE_INTEGER_UNSIGNED:
-					// solution 10-080804-4996
-					// for 64 bit Solaris/AIX (with XlC cplr),
-					// tempVal64 is a signed LONG LONG,
-					// ULONG_MAX is unsigned LONG of 'FFFFFFF....',
-					// somehow it will evaluate tempVal64 GT ULONG_MAX
-					// so cast the tempVal64 to (ULONG)
-					if (tempVal64 < 0)
-					       return IDS_22_003_02;
-					if ((ULONG_P)tempVal64 > ULONG_MAX)
-						return IDS_22_003;
-					ulTmp = (ULONG_P)tempVal64;
-					if  (tempVal64 != ulTmp)
-						retCode = IDS_01_S07;
-					DataPtr = &ulTmp;
-					DataLen = sizeof(ulTmp);
-					break;
-				case SQLTYPECODE_INTEGER:
-					if (tempVal64 < LONG_MIN || tempVal64 > LONG_MAX)
-						return IDS_22_003;
-					lTmp = (LONG)tempVal64;
-					if  (tempVal64 != lTmp)
-						retCode = IDS_01_S07;
-					DataPtr = &lTmp;
-					DataLen = sizeof(lTmp);
-					break;
-				case SQLTYPECODE_LARGEINT:
-				default:
-					DataPtr = &tempVal64;
-					DataLen = sizeof(tempVal64);
-					break;
-				}
 
+                case SQLTYPECODE_SMALLINT_UNSIGNED:
+                    if (tempVal64 < 0)
+                        return IDS_22_003_02;
+                    if ((USHORT)tempVal64 > USHRT_MAX)
```

Comment:
Similar comment to one above. Downcast to (USHORT) doesn't give you what you want.